### PR TITLE
hadoop: depends_on "openjdk@11"

### LINKED ISSUE #107166: Hadoop is broken
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
[william@mbp-2020 logs]% brew config
HOMEBREW_VERSION: 3.5.7-2-g4baff3e
ORIGIN: https://mirrors.ustc.edu.cn/brew.git
HEAD: 4baff3eec111af70599c3fb517576d87ca74e9a2
Last commit: 30 hours ago
Core tap ORIGIN: https://mirrors.ustc.edu.cn/homebrew-core.git
Core tap HEAD: 4c1f447e39fe002aeb712f28925e22d51bea3605
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BOTTLE_DOMAIN: https://mirrors.ustc.edu.cn/homebrew-bottles
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://mirrors.ustc.edu.cn/homebrew-core.git
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 13.1.6 build 1316
Git: 2.32.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.5-x86_64
CLT: N/A
Xcode: 13.4.1

[william@mbp-2020 homebrew-core](master)% brew doctor                                                                               
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Suspicious https://github.com/Homebrew/brew git origin remote found.
The current git origin is:
  https://mirrors.ustc.edu.cn/brew.git

With a non-standard origin, Homebrew won't update properly.
You can solve this by setting the origin remote:
  git -C "/usr/local/Homebrew" remote set-url origin https://github.com/Homebrew/brew
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Start Hadoop HDFS cluster in "Pseudo-Distributed" mode using command ```start-all.sh```

### What happened (include all command output)?

Failed to start node manager and resource manager.
P.S. Logs are located under the ```/usr/local/Cellar/hadoop/3.3.3/libexec/logs```

**hadoop-xxx-nodemanager-xxx.log:**
```
2022-08-03 02:46:38,231 ERROR org.apache.hadoop.yarn.server.nodemanager.NodeManager: Error starting NodeManager
java.lang.ExceptionInInitializerError
	at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
	at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
	at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
	at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
	at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
	at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
	at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
	at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
	at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
	at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
	at com.google.inject.AbstractModule.install(AbstractModule.java:122)
	at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
	at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements.getElements(Elements.java:110)
	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
	at com.google.inject.Guice.createInjector(Guice.java:96)
	at com.google.inject.Guice.createInjector(Guice.java:73)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.build(WebApps.java:417)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:465)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:461)
	at org.apache.hadoop.yarn.server.nodemanager.webapp.WebServer.serviceStart(WebServer.java:125)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
	at org.apache.hadoop.service.CompositeService.serviceStart(CompositeService.java:122)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
	at org.apache.hadoop.yarn.server.nodemanager.NodeManager.initAndStartNodeManager(NodeManager.java:963)
	at org.apache.hadoop.yarn.server.nodemanager.NodeManager.main(NodeManager.java:1042)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @12efa456
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:200)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:194)
	at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
	... 32 more
```

**hadoop-xxx-resourcemanager-xxx.log:**
```
2022-08-03 02:46:35,642 ERROR org.apache.hadoop.yarn.server.resourcemanager.ResourceManager: Error starting ResourceManager
java.lang.ExceptionInInitializerError
	at com.google.inject.internal.cglib.reflect.$FastClassEmitter.<init>(FastClassEmitter.java:67)
	at com.google.inject.internal.cglib.reflect.$FastClass$Generator.generateClass(FastClass.java:72)
	at com.google.inject.internal.cglib.core.$DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at com.google.inject.internal.cglib.core.$AbstractClassGenerator.create(AbstractClassGenerator.java:216)
	at com.google.inject.internal.cglib.reflect.$FastClass$Generator.create(FastClass.java:64)
	at com.google.inject.internal.BytecodeGen.newFastClass(BytecodeGen.java:204)
	at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.<init>(ProviderMethod.java:256)
	at com.google.inject.internal.ProviderMethod.create(ProviderMethod.java:71)
	at com.google.inject.internal.ProviderMethodsModule.createProviderMethod(ProviderMethodsModule.java:275)
	at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:144)
	at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
	at com.google.inject.AbstractModule.install(AbstractModule.java:122)
	at com.google.inject.servlet.ServletModule.configure(ServletModule.java:52)
	at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements.getElements(Elements.java:110)
	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
	at com.google.inject.Guice.createInjector(Guice.java:96)
	at com.google.inject.Guice.createInjector(Guice.java:73)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.build(WebApps.java:417)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:465)
	at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.startWepApp(ResourceManager.java:1389)
	at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.serviceStart(ResourceManager.java:1498)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
	at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.main(ResourceManager.java:1699)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @12efa456
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:200)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:194)
	at com.google.inject.internal.cglib.core.$ReflectUtils$2.run(ReflectUtils.java:56)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at com.google.inject.internal.cglib.core.$ReflectUtils.<clinit>(ReflectUtils.java:46)
	... 29 more
```

### What did you expect to happen?

Node manager and resource manager start successfully.

According to https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+Java+Versions, Hadoop 3.3.x only have java 11 runtime support.  Depending on the latest openjdk is probably a mistake.  Meanwhile, the project should be compiled with openjdk version 8.

BTW, please also build the package with the native libraries, the following warning showed up every time:
```
2022-08-03 22:13:10,101 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
```

### Step-by-step reproduction instructions (by running `brew` commands)
- Install Hadoop
- Setup cluster in "Pseudo-Distributed" mode by following [this guide](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/SingleCluster.html).
- run command ```start-all.sh```
- run command ``` jps```, and managers are not showing in the output

### PR DESCRIPTION
**Pull request by ttc0419**
According to https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+Java+Versions, hadoop only supports java runtime 11 at most.
- Switched dependency from "openjdk" to "openjdk@11" for flume, hadoop and mahout
- Added a test case to check if resource manager server can start successfully with the JDK
- Appended HADOOP_LOG_DIR to env scripts
- Minor refactoring

Closes #107166

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

**Comment by Galaxy0419** [comment]
All tests are passed.  Please review.

File: Formula/hadoop.rb [hunk 1]
```diff
@@ -15,27 +15,49 @@ class Hadoop < Formula
...
+      (libexec/File.basename(path)).write_env_script path, JAVA_HOME: Formula["openjdk@11"].opt_prefix
     end
     # Temporary fix until https://github.com/Homebrew/brew/pull/4512 is fixed
     chmod 0755, Dir["#{libexec}/*.sh"]
   end
 
   test do
     system bin/"hadoop", "fs", "-ls"
+
+    # Test if resource manager can start successfully
```
Comment (SMillerDev) [inline Formula/hadoop.rb:42]: Is there no shell script for this?

Comment (Galaxy0419) [reply Formula/hadoop.rb:42]: The shell script will write log files to the libexec folder, which causes the test always to fail.

Comment (SMillerDev) [reply Formula/hadoop.rb:42]: That should probably be adjusted in the shell script either way. I don't think people would expect to find logs in the install folder. Is there no flag to specify where to log?

Comment (Galaxy0419) [reply Formula/hadoop.rb:42]: I can append ```HADOOP_LOG_DIR``` to env scripts to change the log directory. But I cannot trace the pid to check if the server is alive if I use the script to start the server. The pid file contains the username in its filename, which is different for everyone and the test code would be much more complicated.

File: Formula/hadoop.rb [hunk 2]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
-  depends_on "openjdk"
+  # WARNING: Check https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+Java+Versions before updating JDK version
+  depends_on "openjdk@11"
 
   conflicts_with "yarn", because: "both install `yarn` binaries"
 
   def install
     rm_f Dir["bin/*.cmd", "sbin/*.cmd", "libexec/*.cmd", "etc/hadoop/*.cmd"]
     libexec.install %w[bin sbin libexec share etc]
     Dir["#{libexec}/bin/*"].each do |path|
```
Comment (carlocab) [inline Formula/hadoop.rb:26]:
```suggestion
    (libexec/"bin").each_child do |path|
```
I know this is part of your patch, but may as well fix it while we're here.

File: Formula/hadoop.rb [hunk 3]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
-      (bin/File.basename(path)).write_env_script path, JAVA_HOME: Formula["openjdk"].opt_prefix
+      (bin/File.basename(path)).write_env_script path,
+        JAVA_HOME: Formula["openjdk@11"].opt_prefix, HADOOP_LOG_DIR: "#{var}/hadoop"
     end
     Dir["#{libexec}/sbin/*"].each do |path|
```
Comment (carlocab) [inline Formula/hadoop.rb:30]:
```suggestion
    (libexec/"sbin").each_child do |path|
```

File: Formula/hadoop.rb [hunk 4]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
-      (sbin/File.basename(path)).write_env_script path, JAVA_HOME: Formula["openjdk"].opt_prefix
+      (sbin/File.basename(path)).write_env_script path,
+        JAVA_HOME: Formula["openjdk@11"].opt_prefix, HADOOP_LOG_DIR: "#{var}/hadoop"
     end
     Dir["#{libexec}/libexec/*.sh"].each do |path|
```
Comment (carlocab) [inline Formula/hadoop.rb:34]:
```suggestion
    libexec.glob("libexec/*.sh").each do |path|
```

File: Formula/hadoop.rb [hunk 5]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
+                     "#{libexec}/share/hadoop/common/*:" \
+                     "#{libexec}/share/hadoop/hdfs:" \
+                     "#{libexec}/share/hadoop/hdfs/lib/*:" \
+                     "#{libexec}/share/hadoop/hdfs/*:" \
+                     "#{libexec}/share/hadoop/mapreduce/*:" \
+                     "#{libexec}/share/hadoop/yarn:" \
+                     "#{libexec}/share/hadoop/yarn/lib/*:" \
+                     "#{libexec}/share/hadoop/yarn/*:" \
+                     "#{libexec}/share/hadoop/yarn/timelineservice/*:" \
+                     "#{libexec}/share/hadoop/yarn/timelineservice/lib/*",
```
Comment (carlocab) [inline Formula/hadoop.rb:59]:
```suggestion
    classpaths = %w[
      etc/hadoop
      share/hadoop/common/lib/*
      share/hadoop/common/*
      share/hadoop/hdfs
      share/hadoop/hdfs/lib/*
      share/hadoop/hdfs/*
      share/hadoop/mapreduce/*
      ...
    ].map { |path| libexec/path }
    pid = Process.spawn({
      JAVA_HOME: Language::Java.java_home("11"),
      CLASSPATH: classpaths.join(":"),
```

File: Formula/hadoop.rb [hunk 6]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
+    }, "#{Formula["openjdk@11"].bin}/java", "org.apache.hadoop.yarn.server.resourcemanager.ResourceManager")
+    sleep 8
+    Process.getpgid pid
+    system "curl", "http://127.0.0.1:8088"
```
Comment (carlocab) [inline Formula/hadoop.rb:63]: Can we specify the `8080` so we can select an available port using `free_port`?

File: Formula/mahout.rb [hunk 7]
```diff
@@ -17,10 +17,10 @@ class Mahout < Formula
   end
 
   depends_on "hadoop"
-  depends_on "openjdk"
+  depends_on "openjdk@11"
 
   def install
-    ENV["JAVA_HOME"] = Formula["openjdk"].opt_prefix
+    ENV["JAVA_HOME"] = Formula["openjdk@11"].opt_prefix
```
Comment (carlocab) [inline Formula/mahout.rb:23]:
```suggestion
    ENV["JAVA_HOME"] = Language::Java.java_home("11")
```

File: Formula/flume.rb [hunk 8]
```diff
@@ -11,14 +11,14 @@ class Flume < Formula
...
+  depends_on "openjdk@11"
 
   def install
     rm_f Dir["bin/*.cmd", "bin/*.ps1"]
     libexec.install %w[conf docs lib tools]
     bin.install Dir["bin/*"]
     bin.env_script_all_files libexec/"bin",
-                             JAVA_HOME:  Formula["openjdk"].opt_prefix,
+                             JAVA_HOME:  Formula["openjdk@11"].opt_prefix,
                              FLUME_HOME: libexec
```
[the remark this answers is not in the document]
Comment (Galaxy0419) [reply Formula/flume.rb]: done

File: Formula/hadoop.rb [hunk 9]
```diff
@@ -15,27 +15,51 @@ class Hadoop < Formula
...
     end
     Dir["#{libexec}/sbin/*"].each do |path|
-      (sbin/File.basename(path)).write_env_script path, JAVA_HOME: Formula["openjdk"].opt_prefix
+      (sbin/File.basename(path)).write_env_script path,
+        JAVA_HOME: Formula["openjdk@11"].opt_prefix, HADOOP_LOG_DIR: "#{var}/hadoop"
     end
     Dir["#{libexec}/libexec/*.sh"].each do |path|
-      (libexec/File.basename(path)).write_env_script path, JAVA_HOME: Formula["openjdk"].opt_prefix
+      (libexec/File.basename(path)).write_env_script path,
+        JAVA_HOME: Formula["openjdk@11"].opt_prefix, HADOOP_LOG_DIR: "#{var}/hadoop"
```
[the remark this answers is not in the document]
Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

File: Formula/hadoop.rb (hunk 3 above)
Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

File: Formula/hadoop.rb (hunk 5 above)
Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

File: Formula/mahout.rb (hunk 7 above)
Comment (Galaxy0419) [reply Formula/mahout.rb]: done

File: Formula/hadoop.rb (hunk 4 above)
Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

File: Formula/hadoop.rb (hunk 2 above)
Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

File: Formula/hadoop.rb (hunk 6 above)
Comment (Galaxy0419) [reply Formula/hadoop.rb]: Don't know what ```free_port``` is.  But changed webapp port to 8080.

Comment (carlocab) [reply Formula/hadoop.rb]: Try
```
grep -r free_port "$(brew --repo homebrew/core)/Formula"
```
for many examples.

Comment (Galaxy0419) [reply Formula/hadoop.rb]: done

**Review by carlocab — APPROVED** [review]
Had to split your commits for you in order to enable @BrewTestBot to merge this.

From the formula cookbook:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

Thanks for working on this!

**Pull request closed without merging**